File: src/main/java/com/kelsos/mbrc/commands/model/UpdateNowPlayingList.java
package com.kelsos.mbrc.commands.model;

import com.google.inject.Inject;
import com.kelsos.mbrc.model.MusicTrack;
import com.kelsos.mbrc.interfaces.ICommand;
import com.kelsos.mbrc.interfaces.IEvent;
import com.kelsos.mbrc.model.MainDataModel;
import org.codehaus.jackson.JsonNode;
import org.codehaus.jackson.node.ArrayNode;

import java.util.ArrayList;

public class UpdateNowPlayingList implements ICommand {
    private MainDataModel model;

    @Inject public UpdateNowPlayingList(MainDataModel model) {
        this.model = model;
    }

    @Override public void execute(final IEvent e) {
        ArrayNode node = (ArrayNode) e.getData();
        ArrayList<MusicTrack> playList = new ArrayList<MusicTrack>();
        for (int i = 0; i < node.size(); i++) {
            JsonNode jNode = node.get(i);
            playList.add(new MusicTrack(jNode));
        }

        model.setNowPlayingList(playList);
    }
}

File: app/src/main/kotlin/com/kelsos/mbrc/commands/model/UpdateNowPlayingList.kt
package com.kelsos.mbrc.commands.model

import com.fasterxml.jackson.databind.node.ArrayNode
import com.kelsos.mbrc.data.MusicTrack
import com.kelsos.mbrc.interfaces.ICommand
import com.kelsos.mbrc.interfaces.IEvent
import com.kelsos.mbrc.model.MainDataModel
import rx.Observable
import rx.Subscriber
import rx.schedulers.Schedulers
import timber.log.Timber
import java.util.*
import javax.inject.Inject

class UpdateNowPlayingList
@Inject constructor(private val model: MainDataModel) : ICommand {

  override fun execute(e: IEvent) {

    Observable.create { subscriber: Subscriber<in ArrayList<MusicTrack>> ->
      val node = e.data as ArrayNode
      val playList = ArrayList<MusicTrack>()
      for (i in 0..node.size() - 1) {
        val jNode = node.get(i)
        playList.add(MusicTrack(jNode))
      }
      subscriber.onNext(playList)
      subscriber.onCompleted()
    }.subscribeOn(Schedulers.io()).subscribe({ model.setNowPlayingList(it) }) {
      Timber.e(it, "Failure during now playing parsing")
    }
  }
}
